### ISSUE TITLE
Vulnerability CVE-2019-8320 found in RubyGems

### ISSUE THREAD

**Issue by sergiusignacius** [pact-foundation/pact-net#197]
Hi!

We're using PactNet and our component governance flagged the current version of the library due to its use of Ruby. You can find the CVE [here](https://access.redhat.com/security/cve/cve-2019-8320).

What should be done about it? Is this something easily updated? I could do a PR if I get some instructions.

Thanks!

**Comment by neilcampbell** [comment]
@bethesque Do you know about this one? Any thoughts on this?

**Comment by bethesque** [comment]
> A Directory Traversal issue was discovered in RubyGems 2.7.6 and later through 3.0.2.

$ traveling-ruby-20150715-2.2.2-osx/bin/gem --version
2.4.5

We may be ok because the rubygems version is so old!

**Comment by sergiusignacius** [comment]
Where is this version specified? I'm trying to understand where this flagging of pact comes from...

**Comment by bethesque** [comment]
The underlying Pact Ruby codebase is packaged using Travelling Ruby so that it can be executed without any dependency on Ruby exisiting on the host machine. This allows us to share the executable between languages, and avoid having to reimplement from scratch each time. This shared package is called "pact ruby standalone". You'll find it built into the nuget package somewhere. You can grab the package directly from github here: https://github.com/pact-foundation/pact-ruby-standalone/releases

If you extract the archive, you'll find a bin directory, and inside that, a gem executable. That's what I executed in the above snippet.

**Comment by sergiusignacius** [comment]
Thanks for the explanation. I marked it as a false positive. However, there's a new one:

https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2015-3900

Could we update RubyGems to 2.4.7?

**Comment by sergiusignacius** [comment]
Bump :)

**Comment by bethesque** [comment]
Unfortunately, the package that we use to build the executable, Travelling Ruby, is no longer maintained. I've tried to see if I can get someone to make a new version, but nobody has been up for it so far, and I don't have the time or expertise.

**Comment by sergiusignacius** [comment]
If it's not maintained anymore, is there another way to bundle Ruby in? Would reimplementing it in C# be a choice? From what I understood Travelling Ruby gives us the benefit of not needing to implement pact for every language, but that makes it a very central piece. This latest CVE makes it impossible for us to use Pact, because it is a critical severity security bug. What could we do? I'll gladly help.

**Comment by bethesque** [comment]
There's a bit of back story on this. We'd love to move away from the Travelling Ruby package. Ron Holshausen has written an implementation in Rust, and the plan is to call the Rust functions directly via C bindings from each language, rather than go via a CLI. Unfortunately, this is a lot of work, and it's not something that I (or Ron) have time to do at the moment. Reimplementing Pact (again) is not recommended. It's hard enough to keep the implementations that we have in sync, let alone adding another one. If you're interested in helping out, helping us move the .net implementation over to Rust would be the thing to do.

**Comment by sergiusignacius** [comment]
That does sound better! How could I get started?

- Are there other examples of bindings for another language calling C? 
- How stable is the Rust version? 
- Would we bundle a precompiled binary with the .net DLLs, per arch?

**Comment by bethesque** [comment]
Jump on to slack.pact.io and talk to @uglyog and @mefellows about it.

**Comment by sergiusignacius** [comment]
Closing this one. Given that it's really hard to fix and we didn't find the time to do it, we're turning away from Pact for now.

**Comment by mefellows** [comment]
Thanks @sergiusignacius and apologies - we are on the path to swapping out Ruby for a c-extension, but that will be some time.

FWIW I looked into the issue you described (https://nvd.nist.gov/vuln/detail/CVE-2015-3900). I don't see how this will cause issues. The RubyGems runtime you describe is _not_ actually used in the output package. It's also something you could control via firewall rules if you were paranoid (e.g. blocking requests to unknown hosts).

In other words, I believe there are mitigations.
